(PCP-773) Implement client-side failover for task downloads

ekinanp opened the pull request:
WIP, this is an initial skeleton. I left some TODOs there for a few questions I have about the behavior.

mruzicka commented on lib/src/modules/task.cc:
```diff
@@ -240,30 +239,43 @@ static std::string createUrlEndpoint(const lth_jc::JsonContainer& uri) {
+            // Do nothing here -- we want to try the next master-uri.
+
+            // TODO: Should we collect all the exception messages here to pass in to the overall
+            // Module::ProcessingError if all master_uris fail? Might not be a bad idea.
```
Yes, we should do that.

MikaelSmith replied:
I'd like to distinguish between problems connecting to the server and problems writing the file to disk. That might require a change to Leatherman to make possible (so we can distinguish between types of failures).

ekinanp replied:
Writing to the disk sounds like a filesystem_error, while server stuff is a CURL error. Because I think we really write to the disk as in write to our main file at the last fs::rename part in download_file. But then there's the question of what if things fail at the write_file callback? According to https://curl.haxx.se/libcurl/c/libcurl-errors.html, a CURL-WRITE-ERROR is returned for that case. I think the question is "What would be considered a 'problem writing to disk' vs. connecting to the server?"

MikaelSmith replied:
Ah... I would consider CURL-WRITE-ERROR a 'problem writing to disk'.

ekinanp replied:
What about the rename call at the end?

MikaelSmith replied:
Also filesystem issue. I think our failure modes are: connectivity failure, file operation failure.

ekinanp replied:
Okay, so I'm thinking something along the following lines for the Leatherman changes:

(1) If a filesystem_error is thrown, then that is a file operation failure (writing to disk)

(2) If curl_easy_perform returns CURL_WRITE_ERROR, that is also a file operation failure.

(3) Otherwise, it's a server error.

It seems like (1) and (2) require a new exception class representing file operation failures, unless we want to use fs::filesystem_error. (3) is an http_file_download_exception.

MikaelSmith replied:
👍 probably a new exception is better

mruzicka commented on lib/src/modules/task.cc:
```diff
@@ -240,30 +239,43 @@ static std::string createUrlEndpoint(const lth_jc::JsonContainer& uri) {
 //    (2) If the downloaded file's sha does not match the provided sha, then a PXP
-//        error is returned.
+//        error is returned. TODO: Now that we are trying all the master_uris for
+//        download, should we try another master_uri if the shas do not match?
```
IMO we should.

mruzicka replied:
What about the opposite case: Obtaining a SHA of the most recent version but downloading an older file?
In that case it would make sense to try downloading from other masters precisely for the reason you describe.
But can that happen in a normally operating infrastructure? I guess that depends on the order of the URIs in the `master-uris` configuration parameter.

MikaelSmith replied:
That's fair. In practice, multiple masters will generally be used for MoM/replicas, or load balancers. In the 1st case, the MoM should be up-to-date. With load balancers, you might get an out-of-date master first, then an up-to-date master. I still think it seems odd to retry just because you didn't get the answer you expected.

MikaelSmith replied:
I still don't think we should try unless we get feedback saying that's an issue. It's more complicated and not necessarily the right fix.

MikaelSmith commented:
I wouldn't bother rebasing this, as we'll need to do a clean revert via https://github.com/puppetlabs/pxp-agent/pull/634 before we can merge this.

MikaelSmith commented on lib/src/modules/task.cc:
```diff
@@ -240,30 +239,43 @@ static std::string createUrlEndpoint(const lth_jc::JsonContainer& uri) {
 //    (2) If the downloaded file's sha does not match the provided sha, then a PXP
-//        error is returned.
+//        error is returned. TODO: Now that we are trying all the master_uris for
+//        download, should we try another master_uri if the shas do not match?
```
Not sure I agree. If we get an incorrect sha, presumably that's because a newer version has rolled out. In normally operating infrastructure, we'd expect the certainty of that newer version being rolled out to all masters to increase as time goes on, so I'd expect it to almost always fail on subsequent masters.

MikaelSmith commented:
This needs an update for LTH-142.

ekinanp commented:
Updated the PR with the new leatherman changes. There are still two lingering questions:

(1) In the catch-block for http_file_download exceptions, should I store all of the exception messages and then later print them all out line by line if the entire download fails? 

(2) The ongoing sha discussion.

MikaelSmith commented on lib/src/modules/task.cc:
```diff
@@ -240,30 +239,49 @@ static std::string createUrlEndpoint(const lth_jc::JsonContainer& uri) {
+            // Server-side error, do nothing here -- we want to try the next master-uri.
+
+            // TODO: Should we collect all the exception messages here to pass in to the overall
+            // Module::ProcessingError if all master_uris fail? Might not be a bad idea.
```
I would log something, and throw the last one.

MikaelSmith commented on lib/src/modules/task.cc:
```diff
@@ -240,30 +239,49 @@ static std::string createUrlEndpoint(const lth_jc::JsonContainer& uri) {
-    client.download_file(req, file_path.string());
+            client.download_file(req, file_path.string());
 #endif
+        } catch (lth_curl::http_file_download_exception& e) {
```
That's not true, this needs to exclude http_file_operation_exception.

Also, Leatherman 1.1.2 is tagged, so should be usable.

ekinanp replied:
Thanks for that catch, you're right! Making the fix.

MikaelSmith replied:
Ok, this should be addressed by Leatherman 1.1.3.

MikaelSmith commented on lib/src/modules/task.cc:
```diff
@@ -240,30 +241,54 @@ static std::string createUrlEndpoint(const lth_jc::JsonContainer& uri) {
+        }
+    }
+
+    return result;;
```
Only need 1 semicolon.

MikaelSmith commented on lib/src/modules/task.cc:
```diff
@@ -240,30 +241,54 @@ static std::string createUrlEndpoint(const lth_jc::JsonContainer& uri) {
+// The method returns a tuple (success, err_msg). success is true if the file was downloaded;
+// false otherwise. err_msg contains the most recent http_file_download_exception's error
+// message; it is initially empty.
+static boost::tuple<bool, std::string> downloadTaskFile(const std::vector<std::string>& master_uris,
```
You probably actually want `std::tuple` (that's the header you included.

ekinanp replied:
Will do, I wasn't sure if we should use boost or std there b/c we use boost more often (i.e. not sure of when to use boost vs. std).

MikaelSmith commented on lib/src/modules/task.cc:
```diff
@@ -240,30 +241,54 @@ static std::string createUrlEndpoint(const lth_jc::JsonContainer& uri) {
+    for (auto& master_uri : master_uris) {
+        auto url = master_uri + endpoint;
+        lth_curl::request req(url);
+        // timeout from connection after one minute, can configure
```
I'd make use of the existing `connection-timeout` setting. I believe curl's connection timeout only applies to the initial connection, not a timeout for completing the download.

ekinanp replied:
Yeah, you're right. Their default for the initial connection is five minutes; should we keep that?

MikaelSmith replied:
pxp-agent has a config option `connection-timeout` that defaults to 5 seconds. I think that's what we should use.

MikaelSmith replied:
I guess leave it as-is for now, we'll address that separately.

MikaelSmith commented on lib/src/modules/task.cc:
```diff
@@ -240,30 +241,54 @@ static std::string createUrlEndpoint(const lth_jc::JsonContainer& uri) {
+                                                      const fs::path& file_path,
+                                                      const lth_jc::JsonContainer& uri) {
+    auto endpoint = createUrlEndpoint(uri);
+    boost::tuple<bool, std::string> result = boost::make_tuple(false, "");
```
Use `auto result = std::make_tuple(false, std::string{})`.

ekinanp replied:
Unfortunately I can't, b/c auto makes the result a constant tuple so I can't modify it.

MikaelSmith commented on lib/src/modules/task.cc:
```diff
@@ -240,30 +241,54 @@ static std::string createUrlEndpoint(const lth_jc::JsonContainer& uri) {
+        } catch (lth_curl::http_file_download_exception& e) {
+            // Server-side error, do nothing here -- we want to try the next master-uri.
+            LOG_WARNING("Downloading the task file from the master-uri '{1}' failed. Reason: {2}", master_uri, e.what());
+            boost::get<1>(result) = e.what();
```
With `std::tuple`, you'd use `std::get`.

ekinanp commented:
Alright, updated my PR including the commit message as it's closer to the final state. Did local testing on my machine for the following cases:
  (1) Ensure that pxp-agent falls over to other master-uris.
  (2) Ensure that pxp-agent only throws a download error when all master-uris fail.

The existing functionality from PCP-769 works. Also modified an existing unit test; depending on how CI goes, the modification might have to be turned off for Windows only.

MikaelSmith commented on lib/tests/unit/modules/task_test.cc:
```diff
@@ -305,7 +306,7 @@ TEST_CASE("Modules::Task::executeAction", "[modules][output]") {
     }
 
-    SECTION("errors when a download error occurs and removes the temporary file") {
+    SECTION("if a master-uri has a server-side error, then it proceeds to try the next master-uri. if they all fail, it errors on download and removes all temporary files") {
```
We should also cover the previous error case.

MikaelSmith replied:
I'll add it post-hoc.

MikaelSmith replied:
I guess this test pretty much covers it, just slightly change in the error message.
